File: cpp/isbn-verifier/isbn_verifier.cpp
#include "isbn_verifier.h"

namespace isbn_verifier {

    int count_digits(string isbn) {
        int digits = 0;

        for (auto c : isbn) {
            if (isdigit(c) || tolower(c) == 'x') {
                digits++;
            }
        }

        return digits;
    }

    bool is_valid(string isbn) {
        int sum = 0;
        int digits = count_digits(isbn);

        if (digits < 9 || digits > 10) {
            return false;
        }

        int weight = digits;
        for (auto digit : isbn) {
            char c = tolower(digit);

            if (isdigit(c)) {
                sum += (c - '0') * weight;
            } else if (c == 'x' && weight == 1) {
                sum += 10 * weight;
            } else if (c == '-' || isspace(c)) {
                continue;
            } else {
                return false;
            }

            weight--;
        }

        return sum % 11 == 0;
    }

}  // namespace isbn_verifier

File: lua/isbn-verifier/isbn-verifier.lua
local function is_digit(c)
  return c >= "0" and c <= "9"
end

return {
  valid = function(isbn)
    local sum = 0
    local isbn_len = 0
    local digit_weight = 10

    for i = 1, #isbn do
      local c = string.sub(isbn, i, i)

      if is_digit(c) then
        local digit = tonumber(c, 10)
        sum = sum + digit_weight * digit

        isbn_len = isbn_len + 1
        digit_weight = digit_weight - 1
      elseif i == #isbn and string.upper(c) == "X" then
        sum = sum + digit_weight * 10 

        isbn_len = isbn_len + 1
        digit_weight = digit_weight - 1
      elseif c ~= '-' then
        return false
      end
    end

    return isbn_len == 10 and sum % 11 == 0
  end
}
